source/agora/utils/Log.d

```diff
@@ -49,21 +49,24 @@ public struct Logger
         import core.memory;
         this.logger = Ocean.Log.lookup(moduleName);
         this.logger.buffer(new char[](16384));
-        GC.addRoot(cast(void*)this.logger);
-    }
-
-    /// dtor
-    public ~this ()
-    {
-        import core.memory;
-        GC.removeRoot(cast(void*)this.logger);
     }
 
     public void opDispatch (string call, Args...) (Args args)
     {
         try
         {
-            mixin("this.logger." ~ call ~ "(args);");
+            import core.memory : GC;
+            if (GC.inFinalizer())
+            {
```

Comment:
Please let me know if you know why stderr.writeln doesn't work here, I got no useful message from the compiler. It seems ldc2 is having problems with outputting the real error from iside opDispatch.

![image](https://user-images.githubusercontent.com/19815616/90863056-961a1200-e3b8-11ea-85d5-18d6e4a13683.png)

outputs

![image](https://user-images.githubusercontent.com/19815616/90863087-a5995b00-e3b8-11ea-8a45-78ac27e43e07.png)

Reply:
I think it's a compiler bug. I ran into something like this before and at some point removed a workaround:

https://github.com/bpfkorea/agora/commit/4b60dc4ee703138a79ea342e82086b0fb04d9a15#diff-bea1a77bdfb5d54a9e7d698db79ed2dc

However the underlying problem might still be there.

Reply:
Btw, ideally we wouldn't need this `opDispatch` trampoline. It's only there to make the logger `nothrow`. There should probably be a better way to do this - e.g. just infer nothrow from the call site if possible.

Reply:
> I think it's a compiler bug. I ran into something like this before and at some point removed a workaround:
> 
> [4b60dc4#diff-bea1a77bdfb5d54a9e7d698db79ed2dc](https://github.com/bpfkorea/agora/commit/4b60dc4ee703138a79ea342e82086b0fb04d9a15#diff-bea1a77bdfb5d54a9e7d698db79ed2dc)
> 
> However the underlying problem might still be there.

it is very puzzling because 
std.stdio.writeln("") works, but std.stdio.stderr.writeln("") doesn't

Reply:
anyway reported the issue: https://github.com/ldc-developers/ldc/issues/3545

Reply:
It's actually a DMD front-end issue. LDC uses DMD's front-end. Most compile-time bugs are actually DMD bugs.

Reply:
s/It's/should be

Reply:
Yeah it fails with DMD too:

```diff
diff --git a/dub.json b/dub.json
index 5cb74b88b..7bf021ac5 100644
--- a/dub.json
+++ b/dub.json
@@ -70,7 +70,6 @@
     "buildRequirements": [ "allowWarnings" ],
 
 	"toolchainRequirements": {
-		"dmd": "no",
 		"gdc": "no",
 		"ldc": "~>1.19"
 	},
diff --git a/source/agora/utils/Log.d b/source/agora/utils/Log.d
index e3cd83a6f..22ccecc8a 100644
--- a/source/agora/utils/Log.d
+++ b/source/agora/utils/Log.d
@@ -58,7 +58,7 @@ public struct Logger
             import core.memory : GC;
             if (GC.inFinalizer())
             {
-                writeln("allowing logging from the destructor on a GC thread would risk running into segfaults, please see issue #1128");
+                stderr.writeln("allowing logging from the destructor on a GC thread would risk running into segfaults, please see issue #1128");
                 return;
             }
             else if (logger is null)
@@ -66,7 +66,7 @@ public struct Logger
                 assert(0,"\nplease make sure you are declaring the \nmixin AddLogger!(); statement on top, followed by:\nstatic this{}; followed by:\nstatic ~this{};");
                 return;
             }
-            mixin("this.logger." ~ call ~ "(args);");            
+            mixin("this.logger." ~ call ~ "(args);");
         }
         catch (Exception ex)
         {
```

and:

```
dub build --compiler=dmd
```

Same error.

Reply:
thanks, let me complain there too

Reply:
known issue and marked as critcal... 
https://issues.dlang.org/show_bug.cgi?id=14145

Reply:
Feel free to comment on it. Sometimes commenting helps bring an issue back to focus - there might be compiler devs interested in fixing it.